Authoring Gradle Builds section in Gradle User Manual (#26992)
gradle/gradle: subprojects/docs/src/docs/userguide/authoring-builds/tutorial/partr1_gradle_init.adoc

Enter target version of Java (min. 7) (default: 19):
Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no]
----

NOTE: In this tutorial, Kotlin DSL is used to build a simple Java project (as it is the default DSL starting in Gradle 8.2). All examples are macOS based.

== Step 2. Understanding the Directory layout

The *project root directory* contains all source files from your project.

When you are done with Gradle `init`, the directory should look as follows:

----
.
├── gradle                              // <1>
│   └── wrapper
├── gradlew                             // <2>
├── gradlew.bat                         // <2>
├── settings.gradle.kts                 // <3>
└── app
    ├── build.gradle.kts                // <4>
    └── src
        ├── main
        │   └── java
        │       └── demo
        │           └── App.java        // <5>
        └── test
            └── java
                └── demo
                    └── AppTest.java    // <6>
----
<1> Generated folder for wrapper files
<2> Gradle wrapper start scripts
<3> Settings file to define build name and subprojects

> <4> Build script for `app` dependencies

Reviewer: ```suggestion
<4> Build script for `app` subproject
```

<5> Default Java source folder for `app`

> <6> Default Java test source folder for `app`

Reviewer: ❗ The number icons point to the source files, not the `java` directories.

== Step 3. Review the Gradle Files

The `settings.gradle.kts` file has two interesting lines:

.settings.gradle.kts
[source, kotlin]
----
rootProject.name = "authoring-tutorial"
include("app")
----

- `rootProject.name` assigns a name to the build, which overrides the default behavior of naming the build after the directory it’s in.
- `include("app")` defines that the build consists of one subproject called `app` that contains its own source code and build logic.

More subprojects can be added by additional `include()` statements.

Our build contains one subproject called `app` that represents the Java application we are building.
It is configured in the `app/build.gradle(.kts)` file:

.build.gradle.kts
[source, kotlin]
----
plugins {
    id("application")                                               // <1>
}

repositories {
    mavenCentral()                                                  // <2>
}

dependencies {
    testImplementation("org.junit.jupiter:junit-jupiter:5.9.3")     // <3>
    testRuntimeOnly("org.junit.platform:junit-platform-launcher")

>     implementation("com.google.guava:guava:32.1.1-jre")             // <4>

Reviewer: 🤔 Where does this (or Running Gradle Builds) discuss what `implemetnation` and `testImplementation` mean?  This doesn't attempt to teach this here, so I assume it's meant to be explained elsewhere.

Author: Explained in part 5.

}

> java {                                                              // <5>

Reviewer: 💅 The `<5>` should be on the line with `languageVersion`, I think.

    toolchain {
        languageVersion.set(JavaLanguageVersion.of(19))
    }
}

application {
    mainClass.set("authoring.tutorial.App")                         // <6>
}

tasks.named<Test>("test") {
    useJUnitPlatform()                                              // <7>
}
----
<1> Apply the application plugin to add support for building a CLI application in Java.
<2> Use Maven Central for resolving dependencies.
<3> Use JUnit Jupiter for testing.
<4> This dependency is used by the application.
<5> Define the toolchain version.
<6> Define the main class for the application.
<7> Use JUnit Platform for unit tests.

> The build script in the `app` subproject directory declares the dependencies that the `app` code will need to be assembled, tested, and published.

Reviewer: ```suggestion
The build script in the `app` subproject directory declares the dependencies that the `app` code will need to be assembled and tested.
```

We shouldn't talk about publishing at all at this point, the setup is much more involved.

== Step 4. Review the Code

> The file `app/src/main/java/authoring/tutorial/App.java` contains the main class of the source code:

Reviewer: ```suggestion
The file `app/src/main/java/authoring/tutorial/App.java` contains the main class of the project:
```

.App.java
[source, java]
----
package authoring.tutorial;

public class App {
    public String getGreeting() {
        return "Hello World!";
    }

    public static void main(String[] args) {
        System.out.println(new App().getGreeting());
    }
}
----

The code prints out a simple "Hello World" greeting to the screen.

> The generated test is available at `app/src/test/java/authoring/tutorial/AppTest.java`:

Reviewer: ```suggestion
A test class is available at `app/src/test/java/authoring/tutorial/AppTest.java`:
```

Enter target version of Java (min. 7) (default: 19):

Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no]
----

NOTE: In this tutorial, Kotlin DSL is used to build a simple Java project (as it is the default DSL starting in Gradle 8.2). All examples are macOS based.

== Step 2. Understanding the Directory layout

The *project root directory* contains all source files from your project.

When you are done with Gradle `init`, the directory should look as follows:

----
├── gradle                              // <1>
│   └── wrapper
├── gradlew                             // <2>
├── gradlew.bat                         // <2>
├── settings.gradle.kts                 // <3>
└── app
    ├── build.gradle.kts                // <4>
    └── src
        ├── main
        │   └── java
        │       └── demo
        │           └── App.java        // <5>
        └── test
            └── java
                └── demo
                    └── AppTest.java    // <6>
----
<1> Generated folder for wrapper files
<2> Gradle wrapper start scripts
<3> Settings file to define build name and subprojects
<4> Build script of app project

> <5> Default Java source folder

Reviewer: 💭 Should 5 and 6 also say "for app project?"

<6> Default Java test source folder

> The `tutorial` folder is the *root project directory*.

Reviewer: ```suggestion
The `authoring-tutorial` directory is the *root project directory*.
```

Inside the root project directory are one or more subprojects, build scripts, and the Gradle wrapper.

While the Gradle Wrapper is local to the root project, the Gradle executable is found in the `GRADLE_USER_HOME`.

The `GRADLE_USER_HOME`, which defaults to `USER_HOME/.gradle`, is also where Gradle stores its global configuration properties, initialization scripts, caches, log files and more.

== Step 3. Review the Gradle Files

The `settings.gradle.kts` file has two interesting lines:

.settings.gradle.kts
[source, kotlin]
----
rootProject.name = "authoring-tutorial"
include("app")
----

- `rootProject.name` assigns a name to the build, which overrides the default behavior of naming the build after the directory it’s in.

> - `include("app")` defines that the build consists of one subproject called `app` that contains the source code and build logic.

Reviewer: ```suggestion
- `include("app")` defines that the build consists of one subproject called `app` that contains its own source code and build logic.
```

.build.gradle.kts
[source, kotlin]
----
plugins {
    id("application")                                               // <1>
}

repositories {
    mavenCentral()                                                  // <2>
}

dependencies {
    testImplementation("org.junit.jupiter:junit-jupiter:5.9.3")     // <3>
    testRuntimeOnly("org.junit.platform:junit-platform-launcher")
    implementation("com.google.guava:guava:32.1.1-jre")             // <4>
}

java {                                                              // <5>
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(19))
    }
}

application {
    mainClass.set("authoring.tutorial.App")                         // <6>
}

tasks.named<Test>("test") {
    useJUnitPlatform()                                              // <7>
}
----
<1> Apply the application plugin to add support for building a CLI application in Java.
<2> Use Maven Central for resolving dependencies.
<3> Use JUnit Jupiter for testing.
<4> This dependency is used by the application.
<5> Define the toolchain version.
<6> Define the main class for the application.
<7> Use JUnit Platform for unit tests.

> The build script in the `app` subproject directory includes the dependencies and components the `app` code will need to be assembled, tested, and published.

Reviewer: ```suggestion
The build script in the `app` subproject directory declares the dependencies that the `app` code will need to be assembled, tested, and published.
```

== Step 4. Review the Code

> The file `app/src/main/java/authoring/tutorial/App.java` contains the main class and source code:

Reviewer: ```suggestion
The file `app/src/main/java/authoring/tutorial/App.java` contains the main class' source code:
```

.App.java
[source, java]
----
package tutorial;

public class App {
    public String getGreeting() {
        return "Hello World!";
    }

    public static void main(String[] args) {
        System.out.println(new App().getGreeting());

>     }

Author: Ironically this isn't my code. It's Gradle init!

.AppTest.java
[source, java]
----
package tutorial;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class AppTest {
    @Test void appHasAGreeting() {
        App classUnderTest = new App();
        assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
    }
}
----

The generated test class has a single _JUnit Jupiter_ test.
The test instantiates the `App` class, invokes a method on it, and checks that it returns the expected value.

== Step 5. Run the App

The Application plugin, which was automatically added by `gradle init`, facilitates creating an executable JVM application:

[source, kotlin]
----
plugins {
    id("application")
}
----

Applying the Application plugin implicitly applies the Java plugin and adds tasks like `assemble`, `build`, or `run` to our project.

Thanks to the `application` plugin, you can run the application directly from the command line.
The `run` task tells Gradle to execute the `main` method in the class assigned to the `mainClass` property.

[source]
----
$ ./gradlew run

> Task :app:run
Hello World!

BUILD SUCCESSFUL in 998ms
2 actionable tasks: 2 executed
----

== Step 6. Bundle the App

> The `application` plugin also bundles the application, with all its dependencies, for you.

Author: package*
